### PR TITLE
Implement blog

### LINKED ISSUE #25: Figure out blog infra
From earlier discussion:

```
Routes
/blog = frontpage

    Admin things:
    /new
    /edit/<id>
    /delete
    /delete/<id>

    Everyone things:
    /<id> = view post (with DELETE)
    /archive = view post archive

Blog Pages

    Front Page (x most recent posts)

    Create new post

    View one post (Permalink)
    (admins have edit button)

    List all posts (Archive)
    (admins have delete button)

    Edit existing post

Matt's Idea

/admin/
    has a [New] button
    lists posts
    each post has [Edit --> /admin/edit/<id>] [Delete --> confirmation dialog] buttons

/admin/new
    Unpopulated form for a new post
    Title, date, layout, text

/admin/edit/<id>

```

### PR DESCRIPTION
**Pull request by mplewis**
Closes #25 when we finish everything.

### CODE REVIEW

**Comment by mplewis** [comment]
Let us consider using http://pythonhosted.org/Flask-OpenID/ for admin logins.

**Comment by taylortrimble** [comment]
> Let us consider using http://pythonhosted.org/Flask-OpenID/ for admin logins.

We might as well make authentication a site-wide thing for future architecture, so lets make this its own issue.

**Comment by mplewis** [comment]
Above are some commits I've pushed for the blog admin page. Check them out at /blog/admin, @tylrtrmbl!

File: posts/20130618-newpost.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "title": "Test post, please ignore!",
+    "date": "2013-06-19T05:06:15.860Z",
+    "type": "general",
+    "body": "This is a sample post.\nLorem ipsum dolor sit amet.\nHere's some other cool stuff."
+}
```
Comment (taylortrimble) [inline posts/20130618-newpost.json:6]: Rabble rabble :end:

File: posts/20130622-anotherpost.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "title": "Yet another blog post for friends.",
+    "date": "2013-06-22T01:59:12.501Z",
+    "type": "news",
+    "body": "Here's a sample post. It does some cool stuff!\nCupcake ipsum dolor sit amet.\nReal talk."
+}
```
Comment (taylortrimble) [inline posts/20130622-anotherpost.json:6]: Rabble rabble :end:

File: posts/20130628-secondpost.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+    "title": "The Second Post on the Blog: A Long Title",
+    "date": "2013-06-28T05:11:28.442Z",
+    "type": "general",
+    "body": "This is another sample post!\nBacon ipsum dolor sit amet."
+}
```
Comment (taylortrimble) [inline posts/20130628-secondpost.json:6]: Rabble rabble :end:

File: posts_controller.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        for post_key in self.posts:
+            data = self.posts[post_key]
+        
+            full_file_name = os.path.join(self.posts_dir, post_key + '.json')
+            with open(full_file_name, 'w') as f:
+                json.dump(data, f)
+    
+    def set_posts(self, posts):
+        self.posts = posts
+        self.write_posts()
```
Comment (taylortrimble) [inline posts_controller.py:33]: Rabble rabble :end:

File: static/css/teslaworks.css [hunk 5]
```diff
@@ -0,0 +1,3 @@
+body {
+	margin-top: 40px;
+}
```
Comment (taylortrimble) [inline static/css/teslaworks.css:3]: Rabble rabble :end:

File: templates/blog_list_posts.html [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+</div>
+
+<script>
+$(document).ready(function() {
+    console.log($('.timeago'));
+    $('td.timeago').timeago();
+})
+</script>
+
+{% endblock %}
```
Comment (taylortrimble) [inline templates/blog_list_posts.html:45]: Rabble rabble :end:

File: templates/blog_manage_post.html [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        };
+        input.editor = this;
+        this.update();
+    }
+    var $ = function (id) { return document.getElementById(id); };
+    new Editor($("post-body"), $("preview"));
+</script>
+
+{% endblock %}
```
Comment (taylortrimble) [inline templates/blog_manage_post.html:65]: Rabble rabble :end:

File: posts_controller.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import json
+import os
+
+class PostsController:
```
Comment (taylortrimble) [inline posts_controller.py:4]: Sooooo.... I am sensing a JSON collection controller. That loads a collection of something and populates a dict with the filename's first part as the key.

File: server_dev.py [hunk 9]
```diff
@@ -1,20 +1,56 @@
...
 app = Flask(__name__)
 
+posts_controller = PostsController('posts')
+
 
 @app.errorhandler(404)
 def page_not_found(e):
     return render_template('404.html'), 404
 
+# Front page routes
```
Comment (taylortrimble) [inline server_dev.py:14]: Yay comments for separation.

**Comment by taylortrimble** [comment]
So, I'm officially leaning toward IDs not being hash codes. I don't think they actually need to be, since we can just enforce unique titles. We can auto-generate a title from what's typed in (strip everything no letters, numbers, and spaces, and concatenate with a hyphen)
- "Taylor's Adventure" => "taylors-adventure"
- "Epic &quot;Attack!" => "epic-quotattack"
- "Rails for Charity – Sign up today!" => "rails-for-charity-sign-up-today"

Prettiness in all user URLs ever. And our admins are users. :stuck_out_tongue_winking_eye:

File: server_dev.py [hunk 10]
```diff
@@ -1,20 +1,56 @@
...
 @app.route('/blog')
 def blog():
-    return "Flasktopress isn't quite ready yet, but we're stoked that it's coming."
+    return "Flasktopress isn't coming LOL. Roll your own, son."
+
+@app.route('/blog/<post_id>')
+def view_post(post_id):
+    return "If this worked, you'd be viewing a post with the id " + post_id + '.'
+
+@app.route('/blog/archive/')
```
Comment (taylortrimble) [inline server_dev.py:31]: I've added some notes about routes in #49.

File: server_dev.py [hunk 11]
```diff
@@ -1,20 +1,56 @@
...
+def archive():
+    return "Here's a list of all blog posts! Not really. The archive goes here soon."
+
+@app.route('/blog/admin/')
+def admin_home():
+    blog_posts = posts_controller.get_posts()
+    return render_template('blog_list_posts.html', posts=blog_posts)
```
Comment (taylortrimble) [inline server_dev.py:38]: Could we make this template sound more admin-y? "list_posts" sounds more like what the archive functionality would do.

File: static/js/jquery.timeago.js [hunk 12]
```diff
@@ -0,0 +1,193 @@
+/**
+ * Timeago is a jQuery plugin that makes it easy to support automatically
```
Comment (taylortrimble) [inline static/js/jquery.timeago.js:2]: Is this hosted somewhere externally? Even if we minify everything, we're still going to get big-o benefits from relying on the user's (maybe) cache of popular hosted resources.

File: templates/blog_list_posts.html [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+                <td><a href="#" class="btn btn-danger">Delete</a></td>
+            </tr>    
+        {% endfor %}
+        </tbody>
+    </table>
+</div>
+
+<script>
+$(document).ready(function() {
+    console.log($('.timeago'));
```
Comment (taylortrimble) [inline templates/blog_list_posts.html:40]: Curiosity question: is this something we expect to leave in? Or is this just for debugging now? Or is it debugging we'd like to keep? Or something else.

File: server_dev.py (hunk 10 above)
Comment (taylortrimble) [inline server_dev.py:31]: Let's also be sure to leave off `/` at the end.

File: server_dev.py [hunk 14]
```diff
@@ -1,20 +1,56 @@
...
+
+@app.route('/blog/admin/new/')
```
Comment (taylortrimble) [inline server_dev.py:40]: -- `/`

File: server_dev.py [hunk 15]
```diff
@@ -1,20 +1,56 @@
...
+
+@app.route('/blog/<post_id>')
+def view_post(post_id):
+    return "If this worked, you'd be viewing a post with the id " + post_id + '.'
+
+@app.route('/blog/archive/')
+def archive():
+    return "Here's a list of all blog posts! Not really. The archive goes here soon."
+
+@app.route('/blog/admin/')
```
Comment (taylortrimble) [inline server_dev.py:35]: :facepunch: `/`

**Comment by taylortrimble** [comment]
Let's be sure that the `posts` directory is moved to the `data` directory.

File: server_dev.py (hunk 14 above)
Comment (taylortrimble) [inline server_dev.py:40]: We'll eventually have interfaces for adding new projects, and maybe other "new" things as well. Let's make this `new-post`.

File: server_dev.py [hunk 16]
```diff
@@ -1,20 +1,56 @@
...
+def admin_home():
+    blog_posts = posts_controller.get_posts()
+    return render_template('blog_list_posts.html', posts=blog_posts)
+
+@app.route('/blog/admin/new/')
+def admin_new():
+    return render_template('blog_manage_post.html', post_id=None, post_data=None)
+
+@app.route('/blog/admin/edit/<post_id>')
```
Comment (taylortrimble) [inline server_dev.py:44]: We can edit project info too, so let's do `edit-post`.

File: server_dev.py [hunk 17]
```diff
@@ -1,20 +1,56 @@
...
+def admin_edit(post_id):
+    blog_posts = posts_controller.get_posts()
+    try:
```
Comment (taylortrimble) [inline server_dev.py:47]: One way and only one way – do you want to do it the try/except way and have me change mine below or use "if key in collection else" throughout?

This is in ref to [this line](https://github.com/teslaworksumn/teslaworks.net/blob/master/server_dev.py#L29) in master.

File: templates/blog_list_posts.html [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+                <th>Edit</th>
+                <th>Delete</th>
+            </tr>
+        </thead>
+        <tbody>
+        {% for (post_id, post_data) in posts.iteritems() %}
+            <tr id="{{ post_id }}">
+                <td class="timeago" title="{{ post_data.date }}" />
+                <td title="{{ post_id }}">{{ post_data.title }}</td>
+                <td><a href="#" class="btn btn-info">View</a></td>
```
Comment (taylortrimble) [inline templates/blog_list_posts.html:29]: Do we need `btn-info` and `btn-primary`?

**Comment by taylortrimble** [comment]
Hey! Looked this over and added crap-tons of line comments everywhere. Looks awesome overall.

**Comment by mplewis** [comment]
Fantastic. Are we still looking to scrap the blog in favor of Tumblr? We may be able to use this kind of framework to manage projects instead!

**Comment by taylortrimble** [comment]
Don't ask me why I forgot about that. Just don't. :stuck_out_tongue_winking_eye: Will we be able to host this seamlessly, a-la [punchthrough.com](http://punchthrough.com) with all of our custom branding?

**Comment by mplewis** [comment]
Yup, as long as we can get a hold of the domain servers for teslaworks.net!

**Comment by taylortrimble** [comment]
> Yup, as long as we can get a hold of the domain servers for teslaworks.net!

Sounds like a question for Ben. @bendemaree, what do you think?

**Comment by bendemaree** [comment]
Hey hey, long time no see!  It's obvious from the commits that come to my inbox you guys are working hard.  I apologize for not being around so much, but let me know if you'd like any help getting this running; I can put it some time now.  I heart Flask a lot, also.

In response to your question: so long as we have the registrar login (and we do) then we're all set.

**Comment by taylortrimble** [comment]
Circling back again, I just remembered a super important goal in all of this.

One thing we said earlier is that we wanted blog pushes to be equivalent to email pushes (by interacting with the MailChimp API). This is still a priority of mine, and I want to know if it's a possibility within the Tumblr ecosystem.

Another important consideration: One of our motivating factors for moving to Tumblr was simply resource hosting - however, keep in mind that while this is difficult to tackle we need to do it anyway for the project thumbnails and photos.

If we can get file upload working anyway, the blog/news system might still be best in-house. Rudimentary is fiiine.

I know you won't want to do this :stuck_out_tongue_winking_eye: but lemme know your thoughts!

**Comment by mplewis** [comment]
The main problem I have with implementing the blog in-house is that we have to reinvent the wheel when it comes to writing a blog editor and a post rendering engine.

Ideally, TeslaWorks blog posts should be dripping with media -- photos of cool stuff we did or a video of that cool stuff in action. Right?

It's really easy to write blog software that handles text, but it's a lot harder to make that same software handle:
- Videos
- Photos
- Other multimedia

I think our time would be best spent by outsourcing the blog software to a platform like Tumblr or even Wordpress. Blogging with multimedia is, in my mind, a solved problem, and it's made easier by the fact that our writers will most likely be familiar with the interface of Tumblr. This makes blogging with multimedia easy and fun!

If we do decide to go about this route, we also benefit from having one less point of entry into the site. Yay security.

---

Regarding MailChimp: I liked this idea a lot at first -- post now, email immediately. However, I think that the body of an email should be different from that of a blog post. IMO an email about TeslaWorks news should be more of a short teaser, and a news post should be comprehensive and written like a news article.

For an example of what I think our email newsletters might sound like, read [this sales pitch](http://pastebin.com/4fxbLTPE). This one is selling a product, and we're selling a post. If we can get that idea -- sell the reader on the blog post and get them to the site -- into our emails, I think it would help drive readers to our site.

tl;dr I feel that the email newsletter sales pitch is fundamentally different from even the blog post teaser paragraph. Thoughts?

---

I believe resource hosting is important as well. Are we planning on constructing a web editor for projects, since we don't want users editing JSON directly? If so, we could handle image uploads, storage, etc. at that point alongside the project dashboard.

---

Let me know what you think!

**Comment by taylortrimble** [comment]
I've been putting this off since it's a lot to wrap my head around, but here's my shot:

I don't think that I had planned for blog posts that were long and in-depth, which is why the blog-to-email thing works better in my head than it does in real life. In the interest of moving forward (which I know you know you could anyway, we've both just been working on other things) let's assume a traditional model of blog-to-the-side and email-separate for the first deploy. We have plenty of other integrations to make anyway.

To address popular/share worthy posts though, let's add an issue for adding special teslaworks.net/imma-special-link bounces.

Project web editor yes! One big part that seems small inside the head, but a crucial element.

**Comment by taylortrimble** [comment]
Nuking this puppy. Tumblr onward!
